app_check/src/swig/app_check.i

```diff
@@ -20,14 +20,187 @@
 #include "app_check/src/include/firebase/app_check/app_attest_provider.h"
 #include "app_check/src/include/firebase/app_check/debug_provider.h"
 #include "app_check/src/include/firebase/app_check/device_check_provider.h"
 #include "app_check/src/include/firebase/app_check/play_integrity_provider.h"
-%}
+
+namespace firebase {
+namespace app_check {
+
+typedef void (SWIGSTDCALL *GetTokenFromCSharp)(const char* app_name, int key);
+typedef void (SWIGSTDCALL *TokenChanged)(const char* app_name, AppCheckToken* token);
+
+static GetTokenFromCSharp g_get_token_from_csharp = nullptr;
+static int g_pending_token_keys = 0;
+static std::map<int, std::function<void(AppCheckToken, int, const std::string&)>> g_pending_get_tokens;
+
+static TokenChanged g_token_changed = nullptr;
+
+// Called from C# when a token is fetched, to callback into the C++ SDK.
+void FinishGetTokenCallback(int key, const char* token, int64_t expire_ms,
+                            int error_code, const char* error_message) {
+  // Get the function from the map, and erase it
+  auto callback = g_pending_get_tokens[key];
+  g_pending_get_tokens.erase(key);
+
+  AppCheckToken app_check_token;
+  app_check_token.token = token;
+  app_check_token.expire_time_millis = expire_ms;
+  callback(app_check_token, error_code, error_message);
+}
+
+// Wrapper that calls g_get_token_from_csharp. Should be used with the
+// callback logic to guarantee it is on the Unity thread.
+static void CallGetTokenFromCSharp(int key, const char* name) {
+  if (g_get_token_from_csharp) {
+    g_get_token_from_csharp(name, key);
+  } else {
+    // The C# callback has disappeared, so fail the C++ call.
+    FinishGetTokenCallback(key, "", 0, kAppCheckErrorInvalidConfiguration,
+      "Missing AppCheckProvider C# configuration");
+  }
+}
+
+// C++ implementation of the AppCheckProvider that calls up to the
+// C# library.
+class SwigAppCheckProvider : public AppCheckProvider {
+ public:
+  SwigAppCheckProvider(App* app)
+    : app_(app) {}
+
+  ~SwigAppCheckProvider() override {}
+
+  void GetToken(std::function<void(AppCheckToken, int, const std::string&)>
+                  completion_callback) override {
+    if (g_get_token_from_csharp) {
+      // Save the callback in the map, and generate a key
+      int key = g_pending_token_keys++;
+      g_pending_get_tokens[key] = completion_callback;
+      // Queue a call to the C# function that will generate the token.
+      firebase::callback::AddCallback(
+        new firebase::callback::CallbackValue1String1<int>(
+          key, app_->name(), CallGetTokenFromCSharp));
+    } else {
+      completion_callback({}, kAppCheckErrorInvalidConfiguration,
+        "Missing AppCheckProvider C# configuration");
+    }
+  }
+
+ private:
+  App* app_;
+};
+
+// C++ implementation of the AppCheckProviderFactory that can
+// forward to the C# implementation.
+class SwigAppCheckProviderFactory : public AppCheckProviderFactory {
```

Comment:
The way that these C++ and C# factories and providers interact was not intuitive to me.
I think I am starting to understand.

For iOS and Android the pattern that I followed was native provider and factory wrapping C++ provider and factory one to one. So native factory's CreateProvider calls C++ factory's CreateProvider and then creates a native provider to wrap the newly created C++ provider.

If I understand correctly, there is a single C++ SwigAppCheckProviderFactory which has no reference to the C# factories. It creates one SwigAppCheckProvider for each App, but these C++ providers have no reference to C# providers. Instead GetToken for these providers calls into a single static g_get_token_from_csharp method which looks up C# providers by name.

The way that the C# providers are used is that there is a map from app name to C# provider, and when the GetToken is called it finds the corresponding C# provider in the map.
The way that the C# factories work are used is that when the static C# factory changes, the provider map is cleared. When it next tries to lookup providers in that map it will use the new C# factory to create a new C# provider.

I think this pattern is probably fine. I'm assuming that the benefit of this pattern is a smaller number of classes and methods that call back and forth through the C++ and C# layers.

It will probably be good to add some comments to clarify the responsibility of each of these classes to help future readers of the code.

For example
SwigAppCheckProviderFactory
// This C++ factory is independent of C# factories. The current C# factory is stored in the variable appCheckFactory.

SwigAppCheckProvider
// This C++ provider is independent of C# providers. Logic to select the corresponding C# provider happens later in GetTokenFromCSharpMethod.

GetTokenFromCSharpMethod
// Given an app, create a C# provider if one does not already exist, and call GetToken on the C# provider.

Reply:
Yeah, because the C++/C# layer can be a bit tricky, especially with the memory management, it is following that pattern of setting a single generic factory, which passes everything to C# to then figure out.  I updated the comments to better reflect that, let me know if that helps clarify things.